File: problem-of-the-day/day06/Linked List Cycle II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None





class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        slow = head
        fast = head
        start = head

        if head is None:
            return None

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if fast == slow:
                while slow != start:
                    slow = slow.next
                    start = start.next
                return slow

        return None

File: problem-of-the-day/day06/Linked List Cycle II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode*slow = head;
        ListNode*fast = head;
        while(fast && fast->next){
            slow = slow->next;
            fast = fast->next->next;
            if(slow==fast)break;
        }
        if(!fast || !(fast->next))return nullptr;
        ListNode*home = head;
        while(home!=slow){
            slow=slow->next;
            home=home->next;
        }
        return home;
    }
};